### PR TITLE
Implement IpmiPowerRestorePolicy module

### PR DESCRIPTION
**Pull request by MarcChen46**
# Preface

Please ensure you have read the [contribution docs](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md) prior
to submitting the pull request. In particular,
[pull request guidelines](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md#pull-request-best-practices).

## Description

Implement IpmiPowerRestorePolicy module for configuring Power Restore Policy via IPMI Chassis Command

For each item, place an "x" in between `[` and `]` if true. Example: `[x]`.
_(you can also check items in the GitHub UI)_

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Implement IpmiPowerRestorePolicy module for configuring Power Restore Policy via IPMI Chassis Command, it provides a PlatformPowerRestorePolicyConfigurationLib to retrieve platform setting.
- [x] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Unit tests
- [x] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: add feature readme file

## How This Was Tested

1. Verified the build by pipeline.
2. Verified the Power Restore Policy can be configured correctly on real HW.

## Integration Instructions

To leverage this feature,

1. Add the following to your platform DSC's PEI section:

    ```ini

    IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.inf

    ```

2. Add the following to your platform FDF's PEI section:

    ```ini

    INF  IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.inf

### CODE REVIEW

**Comment by MarcChen46** [comment]
@cfernald @apop5, please help reviewing this PR when you have time, thanks in advance.

File: IpmiFeaturePkg/Include/Library/PlatformPowerRestorePolicyConfigurationLib.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    PlatformPowerRestorePolicyConfigurationLib.h
+
+    Copyright (c) Microsoft Corporation.
+**/
+
+#ifndef PLATFORM_POWER_RESTORE_POLICY_CONFIGURATION_LIB_H_
+#define PLATFORM_POWER_RESTORE_POLICY_CONFIGURATION_LIB_H_
+
+//
+// Power Restore Policy configuration.
```
Comment (cfernald) [inline IpmiFeaturePkg/Include/Library/PlatformPowerRestorePolicyConfigurationLib.h:12]: Maybe make it clear that these definitions come from the IPMI spec.

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c [hunk 2]
```diff
@@ -0,0 +1,89 @@
+/** @file
+    IpmiPowerRestorePolicy.c
```
[the remark this answers is not in the document]
Comment (MarcChen46) [reply IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c]: updated in https://github.com/microsoft/mu_feature_ipmi/pull/144/commits/4c48d83110b5669392c1f3a70926dc1c133d3cd6

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c [hunk 3]
```diff
@@ -0,0 +1,89 @@
+/** @file
+    IpmiPowerRestorePolicy.c
+
+    Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (MarcChen46) [reply IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c]: updated in https://github.com/microsoft/mu_feature_ipmi/pull/144/commits/4c48d83110b5669392c1f3a70926dc1c133d3cd6

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.inf [hunk 4]
```diff
@@ -0,0 +1,33 @@
+## @file
+#  IpmiPowerRestorePolicy.inf
+#
+#  Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (MarcChen46) [reply IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.inf]: updated in https://github.com/microsoft/mu_feature_ipmi/pull/144/commits/4c48d83110b5669392c1f3a70926dc1c133d3cd6

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/MockApi.c [hunk 5]
```diff
@@ -0,0 +1,66 @@
+/** @file    MockApi.c
+
+  Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (MarcChen46) [reply IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/MockApi.c]: updated in https://github.com/microsoft/mu_feature_ipmi/pull/144/commits/4c48d83110b5669392c1f3a70926dc1c133d3cd6

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/TestIpmiPowerRestorePolicyHost.c [hunk 6]
```diff
@@ -0,0 +1,379 @@
+/** @file    TestIpmiPowerRestorePolicyHost.c
+
+  Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (MarcChen46) [reply IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/TestIpmiPowerRestorePolicyHost.c]: updated in https://github.com/microsoft/mu_feature_ipmi/pull/144/commits/4c48d83110b5669392c1f3a70926dc1c133d3cd6

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/TestIpmiPowerRestorePolicyHost.inf [hunk 7]
```diff
@@ -0,0 +1,34 @@
+## @file    TestIpmiPowerRestorePolicyHost.inf
+#
+#  Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (MarcChen46) [reply IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/TestIpmiPowerRestorePolicyHost.inf]: updated in https://github.com/microsoft/mu_feature_ipmi/pull/144/commits/4c48d83110b5669392c1f3a70926dc1c133d3cd6

File: IpmiFeaturePkg/Library/PlatformPowerRestorePolicyConfigurationLibNull/PlatformPowerRestorePolicyConfigurationLibNull.c [hunk 8]
```diff
@@ -0,0 +1,31 @@
+/** @file
+  PlatformPowerRestorePolicyConfigurationLibNull.h
+
+  Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (MarcChen46) [reply IpmiFeaturePkg/Library/PlatformPowerRestorePolicyConfigurationLibNull/PlatformPowerRestorePolicyConfigurationLibNull.c:4]: updated in https://github.com/microsoft/mu_feature_ipmi/pull/144/commits/4c48d83110b5669392c1f3a70926dc1c133d3cd6

File: IpmiFeaturePkg/Library/PlatformPowerRestorePolicyConfigurationLibNull/PlatformPowerRestorePolicyConfigurationLibNull.inf [hunk 9]
```diff
@@ -0,0 +1,31 @@
+## @file PlatformPowerRestorePolicyConfigurationLibNull.inf
+#
+#  Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (MarcChen46) [reply IpmiFeaturePkg/Library/PlatformPowerRestorePolicyConfigurationLibNull/PlatformPowerRestorePolicyConfigurationLibNull.inf:3]: updated in https://github.com/microsoft/mu_feature_ipmi/pull/144/commits/4c48d83110b5669392c1f3a70926dc1c133d3cd6

File: IpmiFeaturePkg/Include/Library/PlatformPowerRestorePolicyConfigurationLib.h [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+//
+#define  POWER_RESTORE_POLICY_POWER_OFF   0x00
+#define  POWER_RESTORE_POLICY_LAST_STATE  0x01
+#define  POWER_RESTORE_POLICY_POWER_ON    0x02
+#define  POWER_RESTORE_POLICY_NO_CHANGE   0x03  // "Set Power Restore Policy"
```
[the remark this answers is not in the document]
Comment (MarcChen46) [reply IpmiFeaturePkg/Include/Library/PlatformPowerRestorePolicyConfigurationLib.h:18]: updated in https://github.com/microsoft/mu_feature_ipmi/pull/144/commits/4c48d83110b5669392c1f3a70926dc1c133d3cd6

File: IpmiFeaturePkg/Include/Library/PlatformPowerRestorePolicyConfigurationLib.h (hunk 1 above)
Comment (MarcChen46) [reply IpmiFeaturePkg/Include/Library/PlatformPowerRestorePolicyConfigurationLib.h]: updated in https://github.com/microsoft/mu_feature_ipmi/pull/144/commits/4c48d83110b5669392c1f3a70926dc1c133d3cd6

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+#include <PiPei.h>
+#include <Library/DebugLib.h>
+#include <Library/IpmiCommandLib.h>
+#include <Library/PlatformPowerRestorePolicyConfigurationLib.h>
+
+/**
+  Configure Power Restore Policy via IPMI
+
+  @param  FileHandle      Handle of the file
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c:18]:
```suggestion
  @param[in]  FileHandle      Handle of the file

```

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+  @param  PeiServices     List of possible PEI Services.
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c:19]:
```suggestion
  @param[in]  PeiServices     List of possible PEI Services.

```

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+  Status = IpmiGetChassisStatus (&ChassisStatusRespData);
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "[%a] - IpmiGetChassisStatus: %r\n", __FUNCTION__, Status));
+    CurrentPowerRestorePolicy = POWER_RESTORE_POLICY_UNKNOWN;
+  } else {
+    CurrentPowerRestorePolicy = (ChassisStatusRespData.CurrentPowerState >> 5) & 0x3; // [6:5] - power restore policy
+  }
+
+  DEBUG ((DEBUG_INFO, "[%a] - CurrentPowerRestorePolicy: 0x%x\n", __FUNCTION__, CurrentPowerRestorePolicy));
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c:50]:
```suggestion
  DEBUG ((DEBUG_VERBOSE, "[%a] - CurrentPowerRestorePolicy: 0x%x\n", __FUNCTION__, CurrentPowerRestorePolicy));

```

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+  // Get platform power restore policy setting
+  //
+  Status = GetPowerRestorePolicy (&PlatformPowerRestorePolicy);
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "[%a] - GetPowerRestorePolicy: %r\n", __FUNCTION__, Status));
+    return Status;
+  }
+
+  DEBUG ((DEBUG_INFO, "[%a] - PlatformPowerRestorePolicy: 0x%x\n", __FUNCTION__, PlatformPowerRestorePolicy));
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c:62]:
```suggestion
  DEBUG ((DEBUG_VERBOSE, "[%a] - PlatformPowerRestorePolicy: 0x%x\n", __FUNCTION__, PlatformPowerRestorePolicy));

```

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+  //
+  if (PlatformPowerRestorePolicy != POWER_RESTORE_POLICY_NO_CHANGE) {
+    SetRestorePolicyRequest.PowerRestorePolicy.Bits.PowerRestorePolicy = PlatformPowerRestorePolicy;
+    SetRestorePolicyRequest.PowerRestorePolicy.Bits.Reserved           = 0;
+    Status                                                             = IpmiSetPowerRestorePolicy (&SetRestorePolicyRequest, &SetRestorePolicyResponse);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "[%a] - IpmiSetPowerRestorePolicy status: %r\n", __FUNCTION__, Status));
+      return Status;
+    } else if (SetRestorePolicyResponse.CompletionCode != 0) {
+      DEBUG ((DEBUG_INFO, "[%a] - IpmiSetPowerRestorePolicy completion code: %r\n", __FUNCTION__, SetRestorePolicyResponse.CompletionCode));
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c:81]:
```suggestion
      DEBUG ((DEBUG_VERBOSE, "[%a] - IpmiSetPowerRestorePolicy completion code: %r\n", __FUNCTION__, SetRestorePolicyResponse.CompletionCode));

```

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c [hunk 16]
```diff
@@ -0,0 +1,87 @@
+/** @file IpmiPowerRestorePolicy.c
+
+  This file implementing a driver entrypoint that configure Power Restore Policy via IPMI command
+
+  Copyright (c) Microsoft Corporation.
+  PDX-License-Identifier: BSD-2-Clause-Patent
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c:6]:
```suggestion
  SPDX-License-Identifier: BSD-2-Clause-Patent

```

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "[%a] - GetPowerRestorePolicy: %r\n", __FUNCTION__, Status));
+    return Status;
+  }
+
+  DEBUG ((DEBUG_INFO, "[%a] - PlatformPowerRestorePolicy: 0x%x\n", __FUNCTION__, PlatformPowerRestorePolicy));
+
+  if ( PlatformPowerRestorePolicy == CurrentPowerRestorePolicy) {
+    // Just return if CurrentPowerRestorePolicy the same with PlatformPowerRestorePolicy
+    DEBUG ((DEBUG_ERROR, "[%a] - Current Power Restore Policy is consistent with platform setting\n", __FUNCTION__));
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/IpmiPowerRestorePolicy.c:66]:
```suggestion
    DEBUG ((DEBUG_VERBOSE, "[%a] - Current Power Restore Policy is consistent with platform setting\n", __FUNCTION__));

```
Its not really an error, its just information that the system is configured as expected.

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/ReadMe.md [hunk 18]
```diff
@@ -0,0 +1,40 @@
+# IpmiPowerRestorePolicy
+
+## Introduction
+
+The IpmiPowerRestorePolicy configure the Power Restore Policy by IPMI Chassis Command, For command detail, please refer
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/ReadMe.md:5]:
```suggestion
The IpmiPowerRestorePolicy configures the Power Restore Policy by the IPMI Chassis Command. For command details, please refer

```

nit

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/ReadMe.md [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+according to Platform setting via IPMI Chassis command.
+
+## Feature Enablement
+
+To leverage this feature,
+
+1. Add the following to your platform DSC:
+
+    ```ini
+
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/ReadMe.md:21]: should the `[Components.PEIM]` be specified here?

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/TestIpmiPowerRestorePolicyHost.c [hunk 20]
```diff
@@ -0,0 +1,380 @@
+/** @file TestIpmiPowerRestorePolicyHost.c
+
+  Copyright (c) Microsoft Corporation.
+  PDX-License-Identifier: BSD-2-Clause-Patent
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/TestIpmiPowerRestorePolicyHost.c:4]:
```suggestion
  SPDX-License-Identifier: BSD-2-Clause-Patent

```

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/TestIpmiPowerRestorePolicyHost.c [hunk 21]
```diff
@@ -0,0 +1,380 @@
...
+#include <Library/IpmiCommandLib.h>
+#include <Library/PlatformPowerRestorePolicyConfigurationLib.h>
+
+#define UNIT_TEST_NAME     "TestIpmiPowerRestorePolicyHost"
+#define UNIT_TEST_VERSION  "1.0"
+
+//
+// Power Restore Policy configuration.
+//
+#define  POWER_RESTORE_POLICY_POWER_OFF   0x00
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/TestIpmiPowerRestorePolicyHost.c:35]: nit:

Redefinition of definese from PlatformPowerRestorePolicyConfigurationLib.h

File: IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/TestIpmiPowerRestorePolicyHost.inf [hunk 22]
```diff
@@ -0,0 +1,33 @@
+## @file TestIpmiPowerRestorePolicyHost.inf
+#
+#  Copyright (c) Microsoft Corporation.
+#  PDX-License-Identifier: BSD-2-Clause-Patent
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiPowerRestorePolicy/UnitTest/TestIpmiPowerRestorePolicyHost.inf:4]:
```suggestion
#  SPDX-License-Identifier: BSD-2-Clause-Patent

```

File: IpmiFeaturePkg/Library/PlatformPowerRestorePolicyConfigurationLibNull/PlatformPowerRestorePolicyConfigurationLibNull.c [hunk 23]
```diff
@@ -0,0 +1,31 @@
+/** @file PlatformPowerRestorePolicyConfigurationLibNull.c
+
+  Copyright (c) Microsoft Corporation.
+  PDX-License-Identifier: BSD-2-Clause-Patent
```
Comment (apop5) [inline IpmiFeaturePkg/Library/PlatformPowerRestorePolicyConfigurationLibNull/PlatformPowerRestorePolicyConfigurationLibNull.c:4]:
```suggestion
  SPDX-License-Identifier: BSD-2-Clause-Patent

```

**Review by apop5 — APPROVED** [review]

**Review by cfernald — APPROVED** [review]

**Pull request merged**